[flakey] TestEthSubscribe panic: test timed out after 4h0m0s

example: https://github.com/erigontech/erigon/actions/runs/14334315283/job/40177394887

```
panic: test timed out after 4h0m0s
	running tests:
		TestEthSubscribe (3h59m56s)

goroutine 8637 [running]:
testing.(*M).startAlarm.func1()
	testing/testing.go:2373 +0x30c
created by time.goFunc
	time/sleep.go:215 +0x38

goroutine 1 [chan receive, 239 minutes]:
testing.(*T).Run(0x140007924e0, {0x10662a344?, 0x1400194bb48?}, 0x1072fc100)
	testing/testing.go:1751 +0x328
testing.runTests.func1(0x140007924e0)
	testing/testing.go:2168 +0x40
testing.tRunner(0x140007924e0, 0x1400194bc68)
	testing/testing.go:1690 +0xe4
testing.runTests(0x14001793110, {0x1084b12e0, 0x76, 0x76}, {0x4b00000000000000?, 0x4b6c8ccf992e6e04?, 0x108881c40?})
	testing/testing.go:2166 +0x3ac
testing.(*M).Run(0x14001a47720)
	testing/testing.go:2034 +0x588
main.main()
	_testmain.go:281 +0x90

goroutine 8652 [chan receive, 239 minutes]:
github.com/erigontech/erigon/turbo/jsonrpc.TestEthSubscribe(0x14006b96340)
	github.com/erigontech/erigon/turbo/jsonrpc/eth_subscribe_test.go:83 +0x640
testing.tRunner(0x14006b96340, 0x1072fc100)
	testing/testing.go:1690 +0xe4
created by testing.(*T).Run in goroutine 1
	testing/testing.go:1743 +0x314

goroutine 8710 [select, 239 minutes]:
github.com/erigontech/erigon/turbo/privateapi.NewEthBackendServer.func1()
	github.com/erigontech/erigon/turbo/privateapi/ethbackend.go:111 +0xf0
created by github.com/erigontech/erigon/turbo/privateapi.NewEthBackendServer in goroutine 8652
	github.com/erigontech/erigon/turbo/privateapi/ethbackend.go:100 +0x2a8

goroutine 8681 [chan receive, 239 minutes]:
github.com/erigontech/erigon-lib/direct.(*SentryMessagesStreamC).Recv(...)
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/sentry_client.go:228
github.com/erigontech/erigon-lib/direct.(*SentryMessagesStreamC).RecvMsg(0x107326a80?, {0x10717bf00, 0x1400c6ced20})
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/sentry_client.go:238 +0x30
github.com/erigontech/erigon-lib/p2p/sentry.pumpStreamLoop[...]({0x107326a80, 0x140095fc000?}, {0x1097a2d68?, 0x1400289b4b8}, {0x10662d590, 0x11}, 0x14006458f68, 0x1072feb60, 0x1400b9aac10, 0x14007c62200, ...)
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/p2p/sentry/loop.go:143 +0x340
github.com/erigontech/erigon-lib/p2p/sentry.ReconnectAndPumpStreamLoop[...]({0x107326a80, 0x140095fc000}, {0x1097a2d68, 0x1400289b4b8}, 0x14006458f50, {0x10662d590, 0x11}, 0x14006458f68, 0x1072feb60, 0x1400b9aac10, ...)
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/p2p/sentry/loop.go:72 +0x500
github.com/erigontech/erigon/p2p/sentry/sentry_multi_client.(*MultiClient).RecvUploadMessageLoop(0x140056f9340, {0x107326a80, 0x140095fc000}, {0x1097a2d68, 0x1400289b4b8}, 0x14007c62200)
	github.com/erigontech/erigon/p2p/sentry/sentry_multi_client/sentry_multi_client.go:88 +0x16c
created by github.com/erigontech/erigon/turbo/stages/mock.MockWithEverything in goroutine 8652
	github.com/erigontech/erigon/turbo/stages/mock/mock_sentry.go:592 +0x3aa4

goroutine 8685 [select, 239 minutes]:
github.com/erigontech/erigon-lib/p2p/sentry.pumpStreamLoop[...].func2()
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/p2p/sentry/loop.go:122 +0xc4
created by github.com/erigontech/erigon-lib/p2p/sentry.pumpStreamLoop[...] in goroutine 8684
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/p2p/sentry/loop.go:120 +0x1f8

goroutine 8677 [select]:
github.com/erigontech/erigon/turbo/stages/headerdownload.(*HeaderDownload).StartPoSDownloader.func1()
	github.com/erigontech/erigon/turbo/stages/headerdownload/header_algos.go:1368 +0x324
created by github.com/erigontech/erigon/turbo/stages/headerdownload.(*HeaderDownload).StartPoSDownloader in goroutine 8652
	github.com/erigontech/erigon/turbo/stages/headerdownload/header_algos.go:1330 +0xa4

goroutine 8711 [select, 239 minutes]:
github.com/erigontech/erigon-lib/direct.(*SubscribeStreamC).Recv(0x1400ea3a9c0)
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/eth_backend_client.go:107 +0x74
github.com/erigontech/erigon/cmd/rpcdaemon/rpcservices.(*RemoteBackend).Subscribe(0x1400e0ea5f0, {0x1073268c0, 0x1088b8fc0}, 0x14004a74040)
	github.com/erigontech/erigon/cmd/rpcdaemon/rpcservices/eth_backend.go:248 +0x16c
github.com/erigontech/erigon/turbo/rpchelper.New.func1()
	github.com/erigontech/erigon/turbo/rpchelper/filters.go:106 +0x14c
created by github.com/erigontech/erigon/turbo/rpchelper.New in goroutine 8652
	github.com/erigontech/erigon/turbo/rpchelper/filters.go:93 +0x4a0

goroutine 8683 [select, 239 minutes]:
github.com/erigontech/erigon/turbo/stages/mock.(*MockSentry).Messages(0x14007c62008, 0x1057ff80c?, {0x107331ed0, 0x14012771650})
	github.com/erigontech/erigon/turbo/stages/mock/mock_sentry.go:215 +0x1e8
github.com/erigontech/erigon-lib/direct.(*SentryClientDirect).Messages.func1()
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/sentry_client.go:190 +0x6c
created by github.com/erigontech/erigon-lib/direct.(*SentryClientDirect).Messages in goroutine 8681
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/sentry_client.go:188 +0x234

goroutine 8682 [select, 239 minutes]:
github.com/erigontech/erigon-lib/p2p/sentry.pumpStreamLoop[...].func2()
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/p2p/sentry/loop.go:122 +0xc4
created by github.com/erigontech/erigon-lib/p2p/sentry.pumpStreamLoop[...] in goroutine 8681
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/p2p/sentry/loop.go:120 +0x1f8

goroutine 8678 [chan receive, 239 minutes]:
github.com/erigontech/erigon-lib/direct.(*SentryMessagesStreamC).Recv(...)
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/sentry_client.go:228
github.com/erigontech/erigon-lib/direct.(*SentryMessagesStreamC).RecvMsg(0x1400276bd88?, {0x10717bf00, 0x1400e0ea500})
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/sentry_client.go:238 +0x30
github.com/erigontech/erigon-lib/p2p/sentry.pumpStreamLoop[...]({0x107326a80, 0x140095fc000?}, {0x1097a2d68?, 0x1400289b4b8}, {0x106621fb7, 0xb}, 0x14007cf5f68, 0x1072feb60, 0x1400b9aab20, 0x14007c62200, ...)
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/p2p/sentry/loop.go:143 +0x340
github.com/erigontech/erigon-lib/p2p/sentry.ReconnectAndPumpStreamLoop[...]({0x107326a80, 0x140095fc000}, {0x1097a2d68, 0x1400289b4b8}, 0x14006456f50, {0x106621fb7, 0xb}, 0x14006456f68, 0x1072feb60, 0x1400b9aab20, ...)
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/p2p/sentry/loop.go:72 +0x500
github.com/erigontech/erigon/p2p/sentry/sentry_multi_client.(*MultiClient).RecvMessageLoop(0x140056f9340, {0x107326a80, 0x140095fc000}, {0x1097a2d68, 0x1400289b4b8}, 0x14007c62200)
	github.com/erigontech/erigon/p2p/sentry/sentry_multi_client/sentry_multi_client.go:121 +0x1dc
created by github.com/erigontech/erigon/turbo/stages/mock.MockWithEverything in goroutine 8652
	github.com/erigontech/erigon/turbo/stages/mock/mock_sentry.go:589 +0x39c4

goroutine 8679 [select, 239 minutes]:
github.com/erigontech/erigon-lib/p2p/sentry.pumpStreamLoop[...].func2()
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/p2p/sentry/loop.go:122 +0xc4
created by github.com/erigontech/erigon-lib/p2p/sentry.pumpStreamLoop[...] in goroutine 8678
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/p2p/sentry/loop.go:120 +0x1f8

goroutine 8636 [select, 239 minutes]:
github.com/erigontech/erigon/turbo/privateapi.(*EthBackendServer).Subscribe(0x14006ac7490, 0x1057ff80c?, {0x107331cf0, 0x1400ea3a960})
	github.com/erigontech/erigon/turbo/privateapi/ethbackend.go:245 +0x190
github.com/erigontech/erigon-lib/direct.(*EthBackendClientDirect).Subscribe.func1()
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/eth_backend_client.go:73 +0x6c
created by github.com/erigontech/erigon-lib/direct.(*EthBackendClientDirect).Subscribe in goroutine 8711
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/eth_backend_client.go:71 +0xf8

goroutine 8676 [chan receive, 239 minutes]:
github.com/erigontech/erigon/turbo/stages/mock.MockWithEverything.func11()
	github.com/erigontech/erigon/turbo/stages/mock/mock_sentry.go:469 +0x3c
created by github.com/erigontech/erigon/turbo/stages/mock.MockWithEverything in goroutine 8652
	github.com/erigontech/erigon/turbo/stages/mock/mock_sentry.go:468 +0x1bc0

goroutine 8680 [select, 239 minutes]:
github.com/erigontech/erigon/turbo/stages/mock.(*MockSentry).Messages(0x14007c62008, 0x14004951cc0?, {0x107331ed0, 0x14012771470})
	github.com/erigontech/erigon/turbo/stages/mock/mock_sentry.go:215 +0x1e8
github.com/erigontech/erigon-lib/direct.(*SentryClientDirect).Messages.func1()
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/sentry_client.go:190 +0x6c
created by github.com/erigontech/erigon-lib/direct.(*SentryClientDirect).Messages in goroutine 8678
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/sentry_client.go:188 +0x234

goroutine 8686 [select, 239 minutes]:
github.com/erigontech/erigon/turbo/stages/mock.(*MockSentry).Messages(0x14007c62008, 0x14007a20d68?, {0x107331ed0, 0x140127717d0})
	github.com/erigontech/erigon/turbo/stages/mock/mock_sentry.go:215 +0x1e8
github.com/erigontech/erigon-lib/direct.(*SentryClientDirect).Messages.func1()
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/sentry_client.go:190 +0x6c
created by github.com/erigontech/erigon-lib/direct.(*SentryClientDirect).Messages in goroutine 8684
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/sentry_client.go:188 +0x234

goroutine 8712 [select, 239 minutes]:
github.com/erigontech/erigon-lib/direct.(*SubscribeLogsStreamC).Recv(0x14010d102a0)
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/eth_backend_client.go:199 +0x74
github.com/erigontech/erigon/cmd/rpcdaemon/rpcservices.(*RemoteBackend).SubscribeLogs(0x1400e0ea5f0, {0x1073268c0, 0x1088b8fc0}, 0x140034e0060, 0x1400c07a678)
	github.com/erigontech/erigon/cmd/rpcdaemon/rpcservices/eth_backend.go:272 +0x1b4
github.com/erigontech/erigon/turbo/rpchelper.New.func2()
	github.com/erigontech/erigon/turbo/rpchelper/filters.go:134 +0x154
created by github.com/erigontech/erigon/turbo/rpchelper.New in goroutine 8652
	github.com/erigontech/erigon/turbo/rpchelper/filters.go:122 +0x528

goroutine 8691 [select, 239 minutes]:
github.com/erigontech/erigon-lib/direct.(*SubscribeLogsStreamS).Recv(0x14010d10240)
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/eth_backend_client.go:168 +0x74
github.com/erigontech/erigon/turbo/privateapi.(*LogsFilterAggregator).subscribeLogs(0x1400e0ea5a0, {0x107334940, 0x14010d10240})
	github.com/erigontech/erigon/turbo/privateapi/logsfilter.go:149 +0x78
github.com/erigontech/erigon/turbo/privateapi.(*EthBackendServer).SubscribeLogs(0x10661608c?, {0x107334940?, 0x14010d10240?})
	github.com/erigontech/erigon/turbo/privateapi/ethbackend.go:389 +0x2c
github.com/erigontech/erigon-lib/direct.(*EthBackendClientDirect).SubscribeLogs.func1()
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/eth_backend_client.go:135 +0x90
created by github.com/erigontech/erigon-lib/direct.(*EthBackendClientDirect).SubscribeLogs in goroutine 8712
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/eth_backend_client.go:132 +0x118

goroutine 8684 [chan receive, 239 minutes]:
github.com/erigontech/erigon-lib/direct.(*SentryMessagesStreamC).Recv(...)
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/sentry_client.go:228
github.com/erigontech/erigon-lib/direct.(*SentryMessagesStreamC).RecvMsg(0x107326a80?, {0x10717bf00, 0x1400c6cee60})
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/direct/sentry_client.go:238 +0x30
github.com/erigontech/erigon-lib/p2p/sentry.pumpStreamLoop[...]({0x107326a80, 0x140095fc000?}, {0x1097a2d68?, 0x1400289b4b8}, {0x10663e424, 0x18}, 0x14006459f68, 0x1072feb60, 0x1400b9aaca0, 0x14007c62200, ...)
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/p2p/sentry/loop.go:143 +0x340
github.com/erigontech/erigon-lib/p2p/sentry.ReconnectAndPumpStreamLoop[...]({0x107326a80, 0x140095fc000}, {0x1097a2d68, 0x1400289b4b8}, 0x14006459f50, {0x10663e424, 0x18}, 0x14006459f68, 0x1072feb60, 0x1400b9aaca0, ...)
	github.com/erigontech/erigon-lib@v0.0.0-00010101000000-000000000000/p2p/sentry/loop.go:72 +0x500
github.com/erigontech/erigon/p2p/sentry/sentry_multi_client.(*MultiClient).RecvUploadHeadersMessageLoop(0x140056f9340, {0x107326a80, 0x140095fc000}, {0x1097a2d68, 0x1400289b4b8}, 0x14007c62200)
	github.com/erigontech/erigon/p2p/sentry/sentry_multi_client/sentry_multi_client.go:103 +0x134
created by github.com/erigontech/erigon/turbo/stages/mock.MockWithEverything in goroutine 8652
	github.com/erigontech/erigon/turbo/stages/mock/mock_sentry.go:595 +0x3b84
FAIL	github.com/erigontech/erigon/turbo/jsonrpc	14401.213s
```

--- a/rpc/jsonrpc/eth_subscribe_test.go
+++ b/rpc/jsonrpc/eth_subscribe_test.go
@@ -18,7 +18,7 @@ package jsonrpc
 
 import (
 	"context"
-	"fmt"
+	"sync"
 	"testing"
 
 	"github.com/stretchr/testify/require"
@@ -40,7 +40,6 @@ import (
 )
 
 func TestEthSubscribe(t *testing.T) {
-	t.Skip("issue #14546")
 	m, require := mock.Mock(t), require.New(t)
 	chain, err := core.GenerateChain(m.ChainConfig, m.Genesis, m.Engine, m.DB, 7, func(i int, b *core.BlockGen) {
 		b.SetCoinbase(common.Address{1})
@@ -57,14 +56,23 @@ func TestEthSubscribe(t *testing.T) {
 	for _, err = range m.Send(&sentry.InboundMessage{Id: sentry.MessageId_BLOCK_HEADERS_66, Data: b, PeerId: m.PeerId}) {
 		require.NoError(err)
 	}
-	m.ReceiveWg.Wait() // Wait for all messages to be processed before we proceeed
+	m.ReceiveWg.Wait() // Wait for all messages to be processed before we proceed
 
 	ctx := context.Background()
 	logger := log.New()
 	backendServer := privateapi.NewEthBackendServer(ctx, nil, m.DB, m.Notifications, m.BlockReader, logger, builder.NewLatestBlockBuiltStore(), nil)
 	backendClient := direct.NewEthBackendClientDirect(backendServer)
 	backend := rpcservices.NewRemoteBackend(backendClient, m.DB, m.BlockReader)
-	ff := rpchelper.New(ctx, rpchelper.DefaultFiltersConfig, backend, nil, nil, func() {}, m.Log)
+	// Creating a new filter will set up new internal subscription channels actively managed by subscription tasks.
+	// We must wait for the first NEW_SNAPSHOT notification, which is always sent unconditionally by EthBackendServer
+	// at the start of Subscribe, to be sure that the subscription is ready, otherwise we could miss some events.
+	subscriptionReadyWg := sync.WaitGroup{}
+	subscriptionReadyWg.Add(1)
+	onNewSnapshot := func() {
+		subscriptionReadyWg.Done()
+	}
+	ff := rpchelper.New(ctx, rpchelper.DefaultFiltersConfig, backend, nil, nil, onNewSnapshot, m.Log)
+	subscriptionReadyWg.Wait() // This is needed *before* stages.StageLoopIteration, which sends NEW_HEADER events
 
 	newHeads, id := ff.SubscribeNewHeads(16)
 	defer ff.UnsubscribeHeads(id)
@@ -79,7 +87,6 @@ func TestEthSubscribe(t *testing.T) {
 
 	for i := uint64(1); i <= highestSeenHeader; i++ {
 		header := <-newHeads
-		fmt.Printf("Got header %d\n", header.Number.Uint64())
 		require.Equal(i, header.Number.Uint64())
 	}
 }

--- a/turbo/privateapi/ethbackend.go
+++ b/turbo/privateapi/ethbackend.go
@@ -104,7 +104,7 @@ func NewEthBackendServer(ctx context.Context, eth EthBackend, db kv.RwDB, notifi
 		defer func() {
 			if err != nil {
 				if !errors.Is(err, context.Canceled) {
-					logger.Warn("[rpc] terminted subscription to `logs` events", "reason", err)
+					logger.Warn("[rpc] terminated subscription to `logs` events", "reason", err)
 				}
 			}
 		}()
@@ -237,7 +237,7 @@ func (s *EthBackendServer) Subscribe(r *remote.SubscribeRequest, subscribeServer
 	defer func() {
 		if err != nil {
 			if !errors.Is(err, context.Canceled) {
-				s.logger.Warn("[rpc] terminted subscription to `newHeaders` events", "reason", err)
+				s.logger.Warn("[rpc] terminated subscription to `newHeaders` events", "reason", err)
 			}
 		}
 	}()

